pkg/expr/threshold.go

```diff
@@ -0,0 +1,134 @@
+package expr
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"strings"
+
+	"github.com/grafana/grafana/pkg/expr/mathexp"
+)
+
+type ThresholdCommand struct {
+	ReferenceVar  string
+	RefID         string
+	ThresholdFunc string
+	Conditions    []float64
+}
+
+const (
+	ThresholdIsAbove        = "gt"
+	ThresholdIsBelow        = "lt"
+	ThresholdIsWithinRange  = "within_range"
+	ThresholdIsOutsideRange = "outside_range"
+)
+
+func NewThresholdCommand(refID, referenceVar, thresholdFunc string, conditions []float64) (*ThresholdCommand, error) {
+	return &ThresholdCommand{
+		RefID:         refID,
+		ReferenceVar:  referenceVar,
+		ThresholdFunc: thresholdFunc,
+		Conditions:    conditions,
+	}, nil
+}
+
+type ThresholdConditionJSON struct {
+	Evaluator ConditionEvalJSON `json:"evaluator"`
+}
+
+type ConditionEvalJSON struct {
+	Params []float64 `json:"params"`
+	Type   string    `json:"type"` // e.g. "gt"
+}
+
+// UnmarshalResampleCommand creates a ResampleCMD from Grafana's frontend query.
+func UnmarshalThresholdCommand(rn *rawNode) (*ThresholdCommand, error) {
+	rawQuery := rn.Query
+
+	rawExpression, ok := rawQuery["expression"]
+	if !ok {
+		return nil, fmt.Errorf("no variable specified to reference for refId %v", rn.RefID)
+	}
+	referenceVar, ok := rawExpression.(string)
+	if !ok {
+		return nil, fmt.Errorf("expected threshold variable to be a string, got %T for refId %v", rawExpression, rn.RefID)
+	}
+
+	jsonFromM, err := json.Marshal(rawQuery["conditions"])
```

Comment:
How come we need to marshal and then unmarshal?

Reply:
At this point the `rawQuery` has already been deserialized to a `map[string]interface{}` and to unmarshal it to a slice of conditions we need a `[]byte`. I've copied this pattern from another expression (threshold I believe?).

Reply:
What is the type in the `interface{}`? If it's a slice of conditions we can just do a type conversion?